### PR TITLE
Correct Alert Handling for Missing Compression Methods

### LINKED ISSUE #7940: Incorrect alert message for malformed ClientHello
openssl version baba1545105131fa34068f62928322e99d695ab1

When OpenSSL receives a ClientHello message that does not list CompressionMethod.null method it aborts with a decode_error Alert.

RFC 8446 states:
```
      For every TLS 1.3 ClientHello, this vector
      MUST contain exactly one byte, set to zero, which corresponds to
      the "null" compression method in prior versions of TLS.  If a
      TLS 1.3 ClientHello is received with any other value in this
      field, the server MUST abort the handshake with an
      "illegal_parameter" alert.
```

while RFC 5246 states:
```
      This vector MUST contain,
      and all implementations MUST support, CompressionMethod.null.
```

so the correct value would be illegal_parameter

Reproducer: https://github.com/tomato42/tlsfuzzer/pull/489

### PR DESCRIPTION
**Pull request by erbsland-dev**
Fixes #7940: Updated the compression check logic to improve protocol compliance. The code now returns `SSL_AD_DECODE_ERROR` when no compression method is provided in the ClientHello message. It returns `SSL_AD_ILLEGAL_PARAMETER` if the “null” compression method (0x00) is missing.

Additionally, refactored the related test code for enhanced readability and maintainability.

- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
How difficult would it be to add a test?

**Comment by erbsland-dev** [comment]
@t8m 
> How difficult would it be to add a test?

I believe adding a test is crucial for this PR, and I initially committed the code change to ensure that all checks pass. However, I’m having trouble with writing the test. While I see that there’s an existing infrastructure for testing protocols, I’m struggling to understand how it works. Could you point me in the right direction or let me know who I can reach out to for help?

**Comment by tomato42** [comment]
> @t8m
> 
> > How difficult would it be to add a test?
> 
> I believe adding a test is crucial for this PR, and I initially committed the code change to ensure that all checks pass. However, I’m having trouble with writing the test. While I see that there’s an existing infrastructure for testing protocols, I’m struggling to understand how it works. Could you point me in the right direction or let me know who I can reach out to for help?

few of the tlsfuzzer test cases are already automated, so that may be the easiest way to do...

File: ssl/statem/statem_srvr.c [hunk 1]
```diff
@@ -1924,15 +1924,24 @@ static int tls_early_post_process_client_hello(SSL_CONNECTION *s)
         OSSL_TRACE_END(TLS_CIPHER);
     }
 
-    for (loop = 0; loop < clienthello->compressions_len; loop++) {
-        if (clienthello->compressions[loop] == 0)
-            break;
-    }
+    if (SSL_CONNECTION_IS_TLS13(s)) {
+        if (clienthello->compressions_len != 1 ||
```
Comment (tomato42) [inline ssl/statem/statem_srvr.c:1928]: compression methods are defined as `opaque legacy_compression_methods<1..2^8-1>;` which means sending an empty list is a `decode_error`, not an `illegal_parameter`

**Comment by erbsland-dev** [comment]
@tomato42 I’ve implemented a differentiation to report a decoding error (`SSL_AD_DECODE_ERROR`) when no compression methods are specified. Additionally, I’ve introduced a new error code, `SSL_R_TLSV13_ONLY_NULL_COMPRESSION_ALLOWED`, to make this issue more transparent to users of the library.

I’m now working on a method to thoroughly test this strict behaviour.

**Comment by mattcaswell** [comment]
There is already a test for checking the compression method handling with null compression. It would seem straightforward to modify the existing test to check the alert code:

https://github.com/openssl/openssl/blob/6c39d21a4844cab997164454ece9b21186881f2a/test/recipes/70-test_comp.t#L65-L81

**Comment by mattcaswell** [comment]
Here's an example of a test where we are looking for a particular alert code:

https://github.com/openssl/openssl/blob/6c39d21a4844cab997164454ece9b21186881f2a/test/recipes/70-test_tls13hrr.t#L71-L82

https://github.com/openssl/openssl/blob/6c39d21a4844cab997164454ece9b21186881f2a/test/recipes/70-test_tls13hrr.t#L154-L161

**Comment by tomato42** [comment]
you can run the appropriate test by simply adding a section to call test-invalid-compression-methods.py in here: 
https://github.com/openssl/openssl/blob/6c39d21a4844cab997164454ece9b21186881f2a/test/recipes/95-test_external_tlsfuzzer_data/cert.json.in#L29-L34

in the same way as the test-conversation.py is executed...

File: ssl/statem/statem_srvr.c [hunk 2]
```diff
@@ -1924,15 +1923,30 @@ static int tls_early_post_process_client_hello(SSL_CONNECTION *s)
...
+            goto err;
+        }
+    } else {
+        /* <=TLSv1.2: Valid if zero compression is supported. */
+        if (memchr(clienthello->compressions, 0,
+                   clienthello->compressions_len) == NULL) {
+            /* No zero compression method found */
+            SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_NO_COMPRESSION_SPECIFIED);
+            goto err;
+        }
```
Comment (tomato42) [inline ssl/statem/statem_srvr.c:1949]: that's code duplication, isn't it?

wouldn't it be better to fist check if there's at least one value there, irrespective of protocol version, and then if we have negotiated TLS1.3 abort it there's more than one or the one is anything but null?

Comment (erbsland-dev) [reply ssl/statem/statem_srvr.c:1949]: Yes, this is indeed duplicated code, but it was intentional for the sake of readability. The compiler will likely optimize the duplicated code away.

The `if` clause is designed to clearly separate the handling for TLSv1.3+ from the older behaviour. Moving the zero-length check outside of this `if` clause would blur that separation and still require a check for null compression in the array.

Comment (erbsland-dev) [reply ssl/statem/statem_srvr.c:1950]: Here’s the code for comparison:

```c
    /* Verify the compression methods */
    if (clienthello->compressions_len == 0) {
        SSLfatal(s, SSL_AD_DECODE_ERROR,
                 SSL_R_NO_COMPRESSION_SPECIFIED);
        goto err;
    }
    if (SSL_CONNECTION_IS_TLS13(s)) {
        /* >=TLSv1.3: RFC 8446 - only zero compression is allowed */
        if (clienthello->compressions_len > 1) {
            SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER,
                     SSL_R_TLSV13_ONLY_NULL_COMPRESSION_ALLOWED);
            goto err;
        }
        if (clienthello->compressions[0] != 0) {
            SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER,
                     SSL_R_TLSV13_ONLY_NULL_COMPRESSION_ALLOWED);
            goto err;
        }
    } else {
        /* <=TLSv1.2: Valid if zero compression is supported. */
        if (memchr(clienthello->compressions, 0,
                   clienthello->compressions_len) == NULL) {
            /* No zero compression method found */
            SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_NO_COMPRESSION_SPECIFIED);
            goto err;
        }
    }
```

In my opinion, this change doesn’t provide a significant improvement. However, if I misunderstood your original comment, please let me know.

Comment (tomato42) [reply ssl/statem/statem_srvr.c:1950]: the TLS 1.3 ClientHello is by design backwards compatible with earlier versions, so having checks that are universal to all versions is the logical way to implement it

also, for TLS 1.2 we also should send illegal_parameter if the null compression method is missing:
```
      This vector MUST contain,
      and all implementations MUST support, CompressionMethod.null.
      Thus, a client and server will always be able to agree on a
      compression method.
```

Comment (erbsland-dev) [reply ssl/statem/statem_srvr.c:1946]: I’ve updated the code to correctly handle the pre-TLSv1.3 cases:

```c
    /* Verify the compression methods */
    if (clienthello->compressions_len == 0) {
        SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_NO_COMPRESSION_SPECIFIED);
        goto err;
    }
    if (SSL_CONNECTION_IS_TLS13(s)) {
        /* >=TLSv1.3: RFC 8446 - only null compression is allowed */
        if (clienthello->compressions[0] != 0 ||
            clienthello->compressions_len > 1) {
            SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER,
                     SSL_R_TLSV13_ONLY_NULL_COMPRESSION_ALLOWED);
            goto err;
        }
    } else {
        /* <=TLSv1.2: Valid if null compression is supported. */
        if (memchr(clienthello->compressions, 0,
                   clienthello->compressions_len) == NULL) {
            SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER,
                     SSL_R_TLSV1_NULL_COMPRESSION_MISSING);
            goto err;
        }
    }
```

Now, only a zero-length compression method field triggers an `SSL_AD_DECODE_ERROR`, while all other cases trigger an `SSL_AD_ILLEGAL_PARAMETER` error. I also introduced a new error code, `SSL_R_TLSV1_NULL_COMPRESSION_MISSING`, to report the absence of null compression for TLSv1.2 and older protocols.

**Review by tomato42 — CHANGES_REQUESTED** [review]
LGTM, but it's still missing test coverage

**Comment by erbsland-dev** [comment]
@t8m @tomato42 The latest update is adding comprehensive tests and fixing the issue with the wrong alert type for TLSv1.2 when no "null" compression is present.

See also: https://github.com/openssl/openssl/issues/7940#issuecomment-2308783578

**Review by tomato42 — APPROVED** [review]
LGTM

**Review by mattcaswell — APPROVED** [review]
LGTM. Nice job.

**Review by t8m — APPROVED** [review]
Good work! Especially on the tests.

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**